src/basic/in-addr-util.c

```diff
@@ -464,3 +464,44 @@ int in_addr_mask(int family, union in_addr_union *addr, unsigned char prefixlen)
 
         return -EAFNOSUPPORT;
 }
+
+int in_addr_prefix_from_string(const char *p, int *family, union in_addr_union *prefix, uint8_t *prefixlen) {
+        union in_addr_union buffer;
+        const char *e, *l;
+        uint8_t k;
+        int f, r;
+
+        assert(p);
+
+        /* prefixlen */
+        e = strchr(p, '/');
+        if (e) {
+                r = safe_atou8(e + 1, &k);
+                if (r < 0)
+                        return r;
+
+                l = strndupa(p, e - p);
+        } else
+                l = p;
+
+        r = in_addr_from_string_auto(l, &f, &buffer);
+        if (r < 0)
+                return r;
+
+        switch(f) {
+        case AF_INET6:
+                if (k > 128)
+                        return -ERANGE;
+                break;
+        case AF_INET:
+                if (k > 31)
+                        return -ERANGE;
+                break;
```

Comment:
shouldn't we make add some checks here that the bits masked by the prefix length are actually zero?

i.e. 192.168.15.23/24 probably should not be accepted for this, but 192.168.15.0/24 should be...

But I am not entirely sure, there might be usecases to permit non-zero suffixes too, you#d have to check the potential usescases. If there are some uses where a zero suffix is OK and other uses where it is not OK then the function should maybe take an additional flags param that permits selecting which mode of parsing should be used.